File: src/data_rover/validation/validators/type_conformance.py
from __future__ import annotations

import datetime

from ...metamodel.schema import Metamodel
from ..issue import Issue, Severity
from ..scope import Scope


def value_conforms(value, datatype: str, metamodel: Metamodel) -> bool:
    if datatype in metamodel.enums:
        return value in metamodel.enums[datatype]
    if datatype == "string":
        return isinstance(value, str)
    if datatype == "boolean":
        return isinstance(value, bool)
    if datatype == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if datatype == "float":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if datatype == "date":
        return isinstance(value, datetime.date)
    return False


class TypeConformanceValidator:
    def validate(self, model, scope: Scope) -> list[Issue]:
        issues: list[Issue] = []
        mm = model.metamodel
        for el in model.elements.values():
            if not scope.includes(el.id):
                continue
            defs = {p.name: p for p in mm.effective_element_properties(el.type_name)}
            issues.extend(self._check(el.type_name, el.id, defs, el.properties, model))
        for rel in model.relationships.values():
            if not scope.includes(rel.id):
                continue
            defs = {
                p.name: p for p in mm.effective_relationship_properties(rel.type_name)
            }
            issues.extend(
                self._check(rel.type_name, rel.id, defs, rel.properties, model)
            )
        return issues

    def _check(self, type_name, owner_id, defs, properties, model) -> list[Issue]:
        out: list[Issue] = []
        mm = model.metamodel
        for name, value in properties.items():
            pdef = defs.get(name)
            if pdef is None or value is None:
                continue
            values = value if isinstance(value, list) else [value]
            if mm.is_element_type(pdef.datatype):
                for item in values:
                    out.extend(
                        self._reference_issues(
                            type_name, owner_id, name, item, pdef.datatype, model
                        )
                    )
            else:
                for item in values:
                    if not value_conforms(item, pdef.datatype, mm):
                        out.append(
                            Issue(
                                Severity.ERROR,
                                f"{type_name}.{name}: value {item!r} is not a "
                                f"valid {pdef.datatype}",
                                [owner_id],
                            )
                        )
        return out

    def _reference_issues(
        self, type_name, owner_id, prop_name, item, declared, model
    ) -> list[Issue]:
        if not isinstance(item, str):
            return [
                Issue(
                    Severity.ERROR,
                    f"{type_name}.{prop_name}: value {item!r} is not a "
                    f"valid {declared} reference",
                    [owner_id],
                )
            ]
        target = model.elements.get(item)
        if target is None:
            return [
                Issue(
                    Severity.ERROR,
                    f"{type_name}.{prop_name}: reference {item!r} points to "
                    f"no element",
                    [owner_id],
                )
            ]
        mm = model.metamodel
        if target.type_name != declared and not mm.is_element_subtype(
            target.type_name, declared
        ):
            return [
                Issue(
                    Severity.ERROR,
                    f"{type_name}.{prop_name}: reference {item!r} is "
                    f"{target.type_name}, expected {declared} or subtype",
                    [owner_id],
                )
            ]
        return []

File: engine/src/validation/validators/type-conformance.ts
import type { Metamodel } from '../../metamodel/metamodel.ts';
import { getProp, type ElementRec, type Props, type RelRec } from '../../model/records.ts';
import { pyRepr, pyReprValue } from '../../value/repr.ts';
import type { Value } from '../../value/types.ts';
import { errorIssue } from '../issue.ts';
import type { Run, Validator } from '../pipeline.ts';
import { valueConforms } from '../values.ts';

type Def = { datatype: string; isReference: boolean };

/**
 * Every value of a declared property against its datatype; an element-typed
 * property must name an element of that type or a subtype. An entity of a
 * type the metamodel lacks gets that one issue and nothing else.
 */
export class TypeConformance implements Validator {
	readonly checkName = 'type_conformance';
	private readonly mm: Metamodel;
	private readonly elementDefs = new Map<string, Map<string, Def>>();
	private readonly relationshipDefs = new Map<string, Map<string, Def>>();

	constructor(mm: Metamodel) {
		this.mm = mm;
	}

	private defs(typeName: string, ofElement: boolean): Map<string, Def> {
		const cache = ofElement ? this.elementDefs : this.relationshipDefs;
		let defs = cache.get(typeName);
		if (defs === undefined) {
			const mm = this.mm;
			const props = ofElement
				? mm.effectiveElementProperties(typeName)
				: mm.effectiveRelationshipProperties(typeName);
			defs = new Map(
				props.map((p) => [
					p.name,
					{ datatype: p.datatype, isReference: mm.isElementType(p.datatype) }
				])
			);
			cache.set(typeName, defs);
		}
		return defs;
	}

	validateElement(run: Run, el: ElementRec): void {
		if (this.mm.elementType(el.typeName) === undefined) {
			run.out.push(unknownType(el.id, el.typeName));
			return;
		}
		this.check(run, el.typeName, el.id, this.defs(el.typeName, true), el.props);
	}

	validateRelationship(run: Run, rel: RelRec): void {
		if (this.mm.relationshipType(rel.typeName) === undefined) {
			run.out.push(unknownType(rel.id, rel.typeName));
			return;
		}
		this.check(run, rel.typeName, rel.id, this.defs(rel.typeName, false), rel.props);
	}

	private check(
		run: Run,
		typeName: string,
		ownerId: string,
		defs: Map<string, Def>,
		props: Props
	): void {
		for (const name of Object.keys(props)) {
			const def = defs.get(name);
			const value = getProp(props, name)!;
			if (def === undefined || value === null) continue;
			const items = Array.isArray(value) ? value : [value];
			for (const item of items) {
				if (def.isReference) this.reference(run, typeName, ownerId, name, item, def.datatype);
				else if (!valueConforms(item, def.datatype, this.mm)) {
					run.out.push(
						errorIssue(
							`${typeName}.${name}: value ${pyReprValue(item)} is not a valid ${def.datatype}`,
							[ownerId]
						)
					);
				}
			}
		}
	}

	private reference(
		run: Run,
		typeName: string,
		ownerId: string,
		prop: string,
		item: Value,
		declared: string
	): void {
		if (typeof item !== 'string') {
			run.out.push(
				errorIssue(
					`${typeName}.${prop}: value ${pyReprValue(item)} is not a valid ${declared} reference`,
					[ownerId]
				)
			);
			return;
		}
		const target = run.model.findElement(item);
		if (target === undefined) {
			run.out.push(
				errorIssue(
					`${typeName}.${prop}: reference ${pyRepr(item)} points to no element`,
					[ownerId],
					'structural'
				)
			);
			return;
		}
		if (target.typeName !== declared && !this.mm.isElementSubtype(target.typeName, declared)) {
			run.out.push(
				errorIssue(
					`${typeName}.${prop}: reference ${pyRepr(item)} is ${target.typeName}, ` +
						`expected ${declared} or subtype`,
					[ownerId]
				)
			);
		}
	}
}

function unknownType(id: string, typeName: string) {
	return errorIssue(`${id} is an instance of unknown type ${pyRepr(typeName)}`, [id]);
}
